triton/transducer-scorer-backend/src/bls.cc

```diff
@@ -0,0 +1,219 @@
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "bls.h"
+
+namespace triton { namespace backend { namespace scorer {
+
+BLSExecutor::BLSExecutor(TRITONSERVER_Server* server)
+    : server_(server), model_executor_(server)
+{
+}
+
+TRITONSERVER_Error*
+BLSExecutor::PrepareInferenceRequest(
+    TRITONSERVER_InferenceRequest** irequest, const std::string model_name)
+{
+
+  // Create an inference request object. The inference request object
+  // is where we set the name of the model we want to use for
+  // inference and the input tensors.
+  RETURN_IF_ERROR(TRITONSERVER_InferenceRequestNew(
+      irequest, server_, model_name.c_str(), -1 /* model_version */));
+
+  RETURN_IF_ERROR(TRITONSERVER_InferenceRequestSetReleaseCallback(
+      *irequest, InferRequestComplete, nullptr /* request_release_userp */));
+
+  return nullptr;  // success
+}
+
+TRITONSERVER_Error*
+BLSExecutor::PrepareInferenceInput(
+     std::vector<torch::Tensor> & input_tensors, std::vector<const char*> & input_names, TRITONSERVER_InferenceRequest* irequest)
+{
+
+  uint32_t input_count;
+  input_count = input_tensors.size();
+  
+  const char* name;
+  TRITONSERVER_DataType datatype = TRITONSERVER_TYPE_FP16;
+
+  TRITONSERVER_MemoryType data_memory_type = TRITONSERVER_MEMORY_GPU;
+  int64_t data_memory_id = 0; // FIX ME yuekai
+
+  for (size_t count = 0; count < input_count; count++) {
+    name = input_names[count];
+    if(std::strcmp(name,"y")==0){
+      // FIX ME, hard-code for decoder
+      datatype = TRITONSERVER_TYPE_INT64;
+    }
+    std::vector<int64_t> input_shapes;
+    auto shape = input_tensors[count].sizes();
+    for (auto itr = shape.begin(); itr != shape.end(); itr++) {
+        input_shapes.push_back(*itr);
+      }
+    uint32_t dims_count = (uint32_t) input_shapes.size();
+
+    const char* data_buffer = reinterpret_cast<const char *>(input_tensors[count].data_ptr());
```

Comment:
Do you need to check that the tensor is contiguous in memory?

Reply:
No worries. The BackendInputCollector would process all requests tensors and return a contiguous buffer.